pkg/controllers/management/auth/roletemplates/prtb_handler.go

```diff
@@ -0,0 +1,161 @@
+package roletemplates
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	mgmtv3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/rbac"
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/rancher/rancher/pkg/user"
+	crbacv1 "github.com/rancher/wrangler/v3/pkg/generated/controllers/rbac/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type prtbHandler struct {
+	userMGR        user.Manager
+	userController mgmtv3.UserController
+	rtController   mgmtv3.RoleTemplateController
+	crController   crbacv1.ClusterRoleController
+	crbController  crbacv1.ClusterRoleBindingController
+}
+
+func newPRTBHandler(management *config.ManagementContext) *prtbHandler {
+	return &prtbHandler{
+		userMGR:        management.UserManager,
+		userController: management.Wrangler.Mgmt.User(),
+		rtController:   management.Wrangler.Mgmt.RoleTemplate(),
+		crController:   management.Wrangler.RBAC.ClusterRole(),
+		crbController:  management.Wrangler.RBAC.ClusterRoleBinding(),
+	}
+}
+
+// OnChange syncs the required resources used to implement a PRTB. It does the following:
+//   - Create the specified user if it doesn't already exist.
+//   - Create the membership bindings to give access to the cluster.
+//   - Create a binding to the project management role if it exists.
+func (p *prtbHandler) OnChange(_ string, prtb *v3.ProjectRoleTemplateBinding) (*v3.ProjectRoleTemplateBinding, error) {
+	prtb, err := p.reconcileSubject(prtb)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := p.reconcileMembershipBindings(prtb); err != nil {
+		return nil, err
+	}
+
+	return prtb, p.reconcileBindings(prtb)
+}
+
+// OnRemove deletes Cluster Role Bindings that are owned by the PRTB. It also removes the membership binding if no other PRTBs give membership access.
+func (p *prtbHandler) OnRemove(_ string, prtb *v3.ProjectRoleTemplateBinding) (*v3.ProjectRoleTemplateBinding, error) {
+	returnErr := deleteMembershipBinding(prtb, p.crbController)
+
+	ownerLabel := rbac.CreatePRTBOwnerLabel(prtb.Name)
+	currentCRBs, err := p.crbController.List(metav1.ListOptions{LabelSelector: ownerLabel})
+	if err != nil {
+		return nil, errors.Join(returnErr, err)
+	}
+
+	for _, crb := range currentCRBs.Items {
+		errors.Join(returnErr, rbac.DeleteResource(crb.Name, p.crbController))
+	}
+
+	return prtb, returnErr
```

Comment:
`errors.Join` do not modify `returnErr`, please consider this:
```suggestion
	var errs []error
	if err:= deleteMembershipBinding(prtb, p.crbController); err != nil {
		errs = append(errs, err)
	}

	ownerLabel := rbac.CreatePRTBOwnerLabel(prtb.Name)
	currentCRBs, err := p.crbController.List(metav1.ListOptions{LabelSelector: ownerLabel})
	if err != nil {
		errs = append(errs, err)
		return nil, errors.Join(errs...)
	}

	for _, crb := range currentCRBs.Items {
		if err := rbac.DeleteResource(crb.Name, p.crbController); err != nil {
			errs = append(errs, err)			
		}
	}

	return prtb, errors.Join(errs...)
```